File: UnitTestingWorkshop/PhoneNumberFormat.vb
Imports System.Text.RegularExpressions

Public NotInheritable Class PhoneNumberFormat

    Private ReadOnly m_number As String

    Public Sub New(number As String)

        m_number = number
    End Sub


    Public Function FormattedNumber() As String

        If (m_number Is Nothing) Then

            Return String.Empty
        End If
        Return FormatNumber()
    End Function

    Private Function FormatNumber() As String

        Const InternationalCallPrefix As String = "00"
        Const SkipInternationalCallPrefixIndex As Integer = 2
        Const RemoveExtraCharactersPattern As String = "[\s-()]"

        Dim formatted As String = Regex.Replace(m_number, RemoveExtraCharactersPattern, String.Empty)
        If (formatted.StartsWith(InternationalCallPrefix)) Then

            Return "+" + formatted.Substring(SkipInternationalCallPrefixIndex)
        End If
        Return formatted
    End Function

    Public Function ContainsOnlyAllowedCharacters() As Boolean

        If (m_number Is Nothing) Then

            Return True
        End If

        Const SearchLettersPattern As String = "[^\s-()+0-9]+"
        Return Not Regex.IsMatch(m_number, SearchLettersPattern)
    End Function

End Class

File: UnitTestingWorkshopHarjoituksetCSharp/PhoneNumberFormat.cs
using System.Text.RegularExpressions;

namespace UnitTestingWorkshopHarjoituksetCSharp
{
    public sealed class PhoneNumberFormat
    {
        private readonly string m_number;

        public PhoneNumberFormat(string number)
        {
            m_number = number;
        }

        public string FormattedNumber()
        {
            return m_number == null ? string.Empty : FormatNumber();
        }

        private string FormatNumber()
        {
            const string internationalCallPrefix = "00";
            const int skipInternationalCallPrefixIndex = 2;
            const string removeExtraCharactersPattern = @"[\s-()]";

            var formatted = Regex.Replace(m_number, removeExtraCharactersPattern, string.Empty);

            if (formatted.StartsWith(internationalCallPrefix))
            {
                return "+" + formatted.Substring(skipInternationalCallPrefixIndex);
            }

            return formatted;
        }

        public bool ContainsOnlyAllowedCharacters()
        {
            if (m_number == null)
            {
                return true;
            }

            const string searchLettersPattern = @"[^\s-()+0-9]+";
            return !Regex.IsMatch(m_number, searchLettersPattern);
        }
    }
}